First set of refactorings for branding

benthorner opened the pull request:
https://www.pivotaltracker.com/story/show/181415991

Please review commit by commit.

These are intended to make all of the branding code and
tests easier to understand and change as we prepare to
add more functionality on the email branding side.

There's more refactoring to do that builds on this work.
This is just where I've got to so far and a good point to
seek a review - hopefully you agree it's better!

servingUpAces commented on app/main/views/service_settings.py:
```diff
@@ -38,14 +38,17 @@
+    AdminSetEmailBrandingForm,
+    AdminSetLetterBrandingForm,
     BillingDetailsForm,
-    BrandingOptions,
```
BrandingOptions is being referred to on line 93. Which is why the unit tests are failing.

benthorner replied:
Oops, thanks for spotting - it got lost in a rebase.

I've amended the third commit and all the tests are passing now.

idavidmcdonald commented on tests/app/utils/test_branding.py:
```diff
@@ -83,3 +85,48 @@ def test_get_available_choices_with_central_org(
     assert list(options) == expected_options
+
+
+def test_get_available_choices_email_branding_set(
```
I think this test name should be more verbose please. I think it is saying something along the lines of

test_get_available_choices_does_x_when_email_branding_is_already_set?

Apart from I'm not sure what 'x' is? I think maybe some code comments about what behaviour we are seeing/expecting might be useful in these two tests added in this commit please.

benthorner replied:
Sorry I could have been clearer. I've been trying to minimise the diff as I move the tests over; in this case it's exposed how [the original test was also similarly confusing](https://github.com/alphagov/notifications-admin/blob/4903e9d60ef2d31f3cd912697dc94b56b2c9f0dd/tests/app/main/views/test_service_settings.py#L5003).

I've made some changes in https://github.com/alphagov/notifications-admin/pull/4182/commits/acd69b8d7de00afba152ab69954fc240879dec87 that should make it clearer. These tests still need a general coverage overhaul as there are quite a few missing still. Is it clearer now though?

idavidmcdonald reviewed (commented):
Thanks Ben, good small commits as always!

A variety of comments/questions but no major objections to your approach here. In particular, I'm a fan of the choices getting moved into their own utils with unit tests. Keeps the complexity in one smaller place.

benthorner commented on app/main/forms.py:
```diff
@@ -2198,57 +2198,10 @@ def __init__(self, service, branding_type):
-            yield ('organisation', service.organisation.name)
-
-        yield ChooseBrandingForm.FALLBACK_OPTION
+        return (
+            list(branding.get_available_choices(service, branding_type)) +
+            [ChooseBrandingForm.FALLBACK_OPTION]
+        )
```
This is actually an expression, not a tuple: a tuple would have a trailing comma on the end. `get_available_choices` returns a generator, so we can't directly do a `+` on it.

I thought it was a bit clearer to put each operand on its own line given they're both complex expressions; parentheses make that possible but I've also been caught out when I accidentally typed a comma on the end!

benthorner commented on app/main/forms.py:
```diff
@@ -2248,7 +2248,7 @@ def get_available_choices(service, branding_type):
             yield ('organisation', service.organisation.name)
 
-        yield BrandingOptions.FALLBACK_OPTION
+        yield ChooseBrandingForm.FALLBACK_OPTION
```
Sure, fixed in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on app/main/forms.py:
```diff
@@ -2266,6 +2266,16 @@ def validate_something_else(self, field):
 
+class ChooseEmailBrandingForm(ChooseBrandingForm):
+    def __init__(self, service_id):
+        ChooseBrandingForm.__init__(self, service_id, branding_type='email')
```
This is something of a debate in the Python community, due to the unpredictable general behaviour of super. However, we use it everywhere else in this repo, so let's be consistent.

Fixed in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on tests/app/utils/test_branding.py:
```diff
@@ -0,0 +1,28 @@
+    ('emergency_service', None, []),
+    ('other', None, []),
+])
+def test_get_available_choices_no_org(
```
Fair enough - amended in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on tests/app/utils/test_branding.py:
```diff
@@ -0,0 +1,28 @@
+
+
+@pytest.mark.parametrize('branding_type', ['email', 'letter'])
+@pytest.mark.parametrize('org_type, existing_branding, expected_options', [
```
Good spot! This is a relic from an old approach.

Removed in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on tests/app/utils/test_branding.py:
```diff
@@ -26,3 +27,59 @@ def test_get_available_choices_no_org(
+
+
+@pytest.mark.parametrize('branding_type', ['email', 'letter'])
+@pytest.mark.parametrize('org_type, existing_branding, expected_options', [
```
Removed in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on tests/app/utils/test_branding.py:
```diff
@@ -83,3 +85,48 @@ def test_get_available_choices_with_central_org(
+    mocker.patch(
+        'app.organisations_client.get_organisation',
+        return_value=organisation_json()
+    )
```
Turns out the original test was wrong / misleading - it wasn't actually mocking any branding being set. It's debatable if we need this as the behaviour doesn't actually change, even when the test is fixed.

For now, I'm trying to keep the changes simple and will do a proper coverage overhaul of these in a follow-up PR. For now, I've just done the minimum to fix the test in https://github.com/alphagov/notifications-admin/pull/4182/commits/fa3e6435a6fb001bc495b28814fba41f8257c192.

benthorner commented on tests/app/utils/test_branding.py:
```diff
@@ -83,3 +85,48 @@ def test_get_available_choices_with_central_org(
     assert list(options) == expected_options
+
+
+def test_get_available_choices_email_branding_set(
```
Sure, this wasn't a great cutoff point for the PR. While these tests do make sense on their own, here's the rest of the change that also identifies the tests they're based on: https://github.com/alphagov/notifications-admin/pull/4182/commits/e7aeb72fd97c9c6da6c9ad8d9a51e0c1cb1e95de.

Does that help?

benthorner commented on app/main/forms.py:
```diff
@@ -1876,7 +1876,7 @@ def __init__(self, channel, *args, **kwargs):
 
 
-class SetEmailBranding(StripWhitespaceForm):
+class AdminSetEmailBrandingForm(StripWhitespaceForm):
```
> change all our forms that are just shown for platform admins are called 'Admin....Form'

On reflection I think you're right: otherwise I'm introducing a partial convention that will just lead to more confusion when someone sees "Admin..." in one place and no other.

I've applied the new convention to all the relevant forms I could find: https://github.com/alphagov/notifications-admin/pull/4182/commits/f02c2b0b1dc8827eb2451136a7f1c7b2d93ccb9a.

idavidmcdonald replied:
Thanks. Looks good

benthorner commented:
Thanks @idavidmcdonald !

I've had a go at addressing all your comments. The tests of the branding choices still need a bunch of work, but hopefully the extra changes I made are enough to make them understandable for the time being.

idavidmcdonald commented on tests/app/utils/test_branding.py:
```diff
@@ -58,57 +58,39 @@ def test_get_available_choices_service_assigned_to_org(
     ]),
-    ('letter', [
+    ('1234-abcd', [
+        ('govuk', 'GOV.UK'),
```
Ah I understand what this test is doing now based on your commit message. Branding `1234-abcd` is just any generic branding (but not the GOV.UK branding) and since they have a non GOV.UK branding set, we offer them the ability to choose GOV.UK if they so wish. I hadn't clocked that.

Very optional:

1. Change '1234-abcd' to be something like 'some-random-branding' as my initial thoughts were that `1234-abcd` was a special string for our tests and meant something

2. Add a comment next to `('govuk', 'GOV.UK'),` saying because the service has a non GOV.UK branding set we allow them to change back to GOV.UK

idavidmcdonald commented on app/main/forms.py:
```diff
@@ -2180,75 +2180,28 @@ def __init__(self, *args, **kwargs):
-            yield ('organisation', service.organisation.name)
-
-        yield BrandingOptions.FALLBACK_OPTION
+    @classmethod
```
Ah nice swap to classmethod to support `cls.FALLBACK_OPTION`

idavidmcdonald reviewed (approved):
Thanks for the adjustments, looks good!
